pkg/disttask/framework/storage/task_table.go

```diff
@@ -635,42 +654,25 @@ func (stm *TaskManager) IsSchedulerCanceled(ctx context.Context, execID string,
 	return len(rs) == 0, nil
 }
 
-// UpdateFailedSchedulerIDs replace failed scheduler nodes with alive nodes.
-func (stm *TaskManager) UpdateFailedSchedulerIDs(ctx context.Context, taskID int64, replaceNodes map[string]string) error {
-	// skip
-	if len(replaceNodes) == 0 {
+// UpdateSubtasksSchedulerIDs update subtasks' schedulerID.
+func (stm *TaskManager) UpdateSubtasksSchedulerIDs(ctx context.Context, taskID int64, subtasks []*proto.Subtask) error {
+	// skip the update process.
+	if len(subtasks) == 0 {
 		return nil
 	}
-
-	sql := new(strings.Builder)
-	if err := sqlescape.FormatSQL(sql, "update mysql.tidb_background_subtask set state = %? ,exec_id = (case ", proto.TaskStatePending); err != nil {
-		return err
-	}
-	for k, v := range replaceNodes {
-		if err := sqlescape.FormatSQL(sql, "when exec_id = %? then %? ", k, v); err != nil {
-			return err
-		}
-	}
-	if err := sqlescape.FormatSQL(sql, " end) where task_key = %? and state != \"succeed\" and exec_id in (", taskID); err != nil {
-		return err
-	}
-	i := 0
-	for k := range replaceNodes {
-		if i != 0 {
-			if err := sqlescape.FormatSQL(sql, ","); err != nil {
+	err := stm.WithNewTxn(ctx, func(se sessionctx.Context) error {
+		for _, subtask := range subtasks {
+			_, err := ExecSQL(ctx, se,
+				"update mysql.tidb_background_subtask set exec_id = %? where id = %? and state = %?",
```

Comment:
why care about `state`?

maybe we should consider add old exec_id in where clause in case there are 2 owners. ok for now

Reply:
state may changed before updateSubtasks.
The timeline will be like:
1. Dispatcher: getSubtasks
2. scheduler: getSubtasks and mark it as running
3. Dispatcher: rebalance the subtask to other node.